web-components/src/components/barcode-scanner/barcode-scanner.ts

```diff
@@ -0,0 +1,408 @@
+      // clean video
+      this.video.srcObject = null
+      this.stream = null
+    }
+  }
+
+  private sendBarcodeStateEvent(detail: { barcode?: string; state: BarcodeScannerState }) {
+    this.dispatchEvent(new CustomEvent(EventType.BARCODE_SCANNER_STATE, { detail }))
+  }
+
+  private exit(event: Event) {
+    event.stopPropagation()
+    this.stopVideo()
+    this.sendBarcodeStateEvent({
+      state: BarcodeScannerState.EXITED,
+    })
+  }
+
+  private setupFrameCapture() {
+    const video = this.video
+    const processFrame = async () => {
+      if (video.paused || video.ended) {
+        requestAnimationFrame(processFrame)
+        return
+      }
+
+      let imageBitmap: ImageBitmap | undefined
+      try {
+        imageBitmap = await createImageBitmap(video, 0, 0, video.videoWidth, video.videoHeight)
+      } catch (e) {
+        console.error("Error creating image bitmap:", e)
+      }
+
+      if (imageBitmap) {
+        try {
+          const result = await this.detectFn!(imageBitmap)
+          if (result) {
+            this.barcode = result
+            this.sendBarcodeStateEvent({
+              barcode: this.barcode,
+              state: BarcodeScannerState.DETECTED,
+            })
+
+            this.stopVideo()
+            return
+          }
+        } catch (error) {
+          console.error("Barcode detection error:", error)
+        }
+      }
+
+      requestAnimationFrame(processFrame)
+    }
+
+    requestAnimationFrame(processFrame)
+  }
+
+  override disconnectedCallback() {
+    super.disconnectedCallback()
+    this.stopVideo()
+  }
+
+  private renderOverlay() {
+    if (!this.codeReader) {
+      return html` <div class="cta-overlay-wrapper">
+        <div class="centered-overlay">
+          <div class="cta-overlay">
+            <div class="cta-overlay-text">
+              ${msg("Your browser does not support barcode scanning.")}
+            </div>
+          </div>
+        </div>
+      </div>`
+    }
+    if (this.isVideoPlaying) {
+      return html`
+        <div class="cta-overlay-wrapper" @click=${this.askPermission}>
+          <div class="cta-exit-button">
+            <button class="button white-button small" @click=${this.exit}>
+              <close-icon></close-icon><span>${msg("Close")}</span>
+            </button>
+          </div>
+          <div class="centered-overlay">
+            <div class="cta-overlay-frame right-top-border"></div>
+            <div class="cta-overlay-frame left-top-border"></div>
+            <div class="cta-overlay-frame right-bottom-border"></div>
+            <div class="cta-overlay-frame left-bottom-border"></div>
+            <div class="cta-overlay">
+              <div class="cta-overlay-icon">
+                <button class="cta-overlay-button">
+                  <search-icon></search-icon>
+                </button>
+              </div>
+            </div>
+          </div>
+        </div>
+      `
+    }
+    return html`
```

Comment:
Isn't there an in-between state where user have click but we are still setting things up ? (maybe to have a wait message).